### ISSUE TITLE
New Rule: await with arithmetic or function args

### ISSUE THREAD

**Issue by Dahaden** [eslint/eslint#17469]
### Rule details

inline await with arithmetic or function arguments seems to cache values in certain positions for the life of the promise.
This leads to some interesting very unexpected results

### Related ECMAScript feature

Async Await

### What type of rule is this?

Suggests an alternate way of doing something

### Example code

```js
// Async counter
const asyncCounter = (add = 1) => {
    // Return `1` after 0-20 ms
    return new Promise(resolve => setTimeout(resolve, Math.random() * 20, add));

    // Same thing works with a string
    // return new Promise(resolve => setTimeout(resolve, Math.random() * 20, '1'));
}

// Using arithmetic to count with async functions
const arithmeticCounter = async (max) => {
    let badResult = 0;
    const promises = Array(max).fill(0).map(async () => {
        badResult += await asyncCounter();

        // The below have similar effects
        // badResult -= await asyncCounter();
        // badResult = badResult + await asyncCounter();

        // Also break but youll need to change badResult to test
        // badResult *= await asyncCounter(2);
        // badResult /= await asyncCounter(2);
        // badResult %= await asyncCounter(3); // Harder to test but pretty sure it would be the same 

        // This actually works, but should probably be avoided to prevent confusion
        // badResult = await asyncCounter() + badResult;
    });
    await Promise.all(promises);

    console.log(`Arithmetic Bad Result: ${badResult}`);
    // "Arithmetic Bad Result: 1" 

    let goodResult = 0;
    const goodPromises = Array(max).fill(0).map(async () => {
        // Moving the await to its own line fixes this error
        const tempCount = await asyncCounter();
        goodResult += tempCount;
    });
    await Promise.all(goodPromises);

    console.log(`Arithmetic Good Result: ${goodResult}`);
    // "Arithmetic Good Result: 50" 
};

const addFunc = (a, b) => a + b;

const logAndReturnFunc = (a, b, c) => {
    console.log(`A: ${a}, B: ${b}, C: ${c}`);
    return a + b + c;
}

// Example using function instead
const functionCounter = async (max) => {
    let badResult = 0;
    const promises = Array(max).fill(0).map(async () => {
        // Similar to arithmetic but using in function args instead
        badResult = addFunc(badResult, await asyncCounter());

        // This actually works, but should probably be avoided to prevent confusion
        // badResult = addFunc(await asyncCounter(), badResult);
    });
    await Promise.all(promises);

    console.log(`FuncArg Bad Result: ${badResult}`);
    // "FuncArg Bad Result: 1"

    let goodResult = 0;
    const goodPromises = Array(max).fill(0).map(async () => {
        // Again moving the await to its own line fixes this error
        const tempCount = await asyncCounter();
        goodResult = addFunc(goodResult, tempCount);
    });
    await Promise.all(goodPromises);

    console.log(`FuncArg Good Result: ${goodResult}`);
    // "FuncArg Good Result: 50"

    let interestingResult = 0;
    const interestingPromises = Array(max).fill(0).map(async () => {
        
        // Mixing interestingResult before and after `await` for interesting results
        // A is always 0
        // B is always 1
        // C is the current iteration index
        interestingResult = logAndReturnFunc(
            interestingResult,
            await asyncCounter(),
            interestingResult
        );
    });
    await Promise.all(interestingPromises);
};

const main = async () => {
    await arithmeticCounter(50);
    await functionCounter(50);
}

main();
```

### Why should this rule be in the core instead of a plugin?

Async Await have become a huge part of modern ECMAscript 

This bug feels like it has the ability to be easily undetected, or confuse many developers (it confused me and some colleagues!)

I feel like an eslint rule that suggests moving `await` assignments to a new line will prevent this error, and also have a nice side effect of making code cleaner.

### Participation

- [X] I am willing to submit a pull request to implement this rule.

### Additional comments

[Typescript playground for easier testing](https://www.typescriptlang.org/play?#code/PTAEEEGcE8DsGNTwPYFdYBcCmAnAUCrJBqAIYwIDCamuoAvKABSkAmrDoAjAJQMB8oAN55QY0CFAAlLBlQ5YoAAZclZAGbYcoAAwBaAEw7QAW0ijxOWfMWwsAd1AAFHMhMBLSFgA8sVCYAjXH4mK0hkABsANywBUC8MABV3Eyw0DFCscOisABpQAFlSDAALADocUlhWNyY+ACpQI3y2Vh4eAG48CzFJAGVSVNBS91gAc1B7ZBwAa0hJ91KyeIwcUbGeiTArOQVQO0cXN08fP0DgzOyYuITk1PTLyJj8otKKqpqTOtBG5tAAci4-3aXQAvt1JABVSDrMhrUqpDDuRAYZBIGgkeyLEpkCiIdToeBI5BEAgk4hw7GI5HUdBaTjkOCIJgmUgADwAXPt-EEcHx6IIROJQBFZKAAmwZJBUBESIwdF1hYQKQAHVweLzzRjgHCVaAs9k8MrqdwRCJMHRG1kqlh45j8wWbYUS1hSmUkADUjFI9lIi1xTNptBwdS6TvEkkSJViQQiyEcJVI1xhHgipG0WHU6iwRPMwuFkhdbtloD03t9-sZVAxuFD4d6YCLWXdnCb0pLHrIFZIVfgQa0dfrWwgEXC4qspBm4tQJGgaDN+ywWA4qKQifGMclzZLq+wxCHha37ZI9XLfp7eP7tYMnQPjaPLeAZ8rl5rIZvivzDfFD5LAFJnwvQM3yYABmTphwACXTVg6F3LISACGdQDVWQMGgeJ5Fif0phlDggmGaN4kGWJui-b8o08MgiVQUgzQwqZZkgfIkJISASnnDg1WQCUAgiDCCKTZB3Fglc0VQmJMHRWACRhEk7x-V1tzlLtzwDas6VrPhOzbd1PzEUFb2FH01KODUsDKOjzW48zIFvTZlUiCy4zGJglB1KlZGRUAACE2GkZSuQAEiEXTZVBJQjIjMAACIPIRLzED8jhiwwLkuBi0AyPEUUSDGZBkCU49OAVBzyTygrWDMk4tQgXVSH1Vk2SNE0zQtK1SBtXt7TiIVyMkApkCiWFSliEz-VXRZ5njRQIlGWITTZLJCKo3BXHwcjpIpbATBVfsGW7dS+xAqKv3ywrUtAL1hiwXb+300BDIe8aSGqrxLLa86qvVGr7KVcknLKFy3PikpqUQABxSqAuPYKhC+1KItO4c4vhMHEtAKHCph90uQAVh0TK8FBMNlR7dgADFCU4FguTOXkWLpnlcAdZYdNJ8qRWQMZwGqGRdlgKmEBp0gmfOHBGe5cX8ngMXeVZvqxEc0Uge5kG4dIUF8h8uGAi10BKDh+AkYenYbDZ8UrqQMEITAABRNlBhVUVQFQGFxlAAkEGJRRRmILA2DJIgSC9ol3BJK9tG9O0DU5KX5d6zZcsUy75QesmUJ+zVOB1PVY5a01zUtMprVtJkeoFYQFL6FJTXTYY0XTTykUQVjXfdiZRk9wkfbhMZ5j97BA82sKVNaIX4CYUeWkO3tI7qP7+rASj5lIGirIY6Y5hY5D2M4zOeNIPiBLGqJhNEhvM6wSSSEIWTw9gBTR4ZSnCRYWfX00kMeBY38MGRp6mwXrOCzhZKyTAbK-TDP9IggNgZKAnjqCYyUcayl1n-E2mxJAxUQTgZB-lUrpRitlMQycEbKRKunTmX03rLW1PVRqhpjSF3aiXTqZdhbfErorfMkhwBjD9IoEwQ0RpEWAZNDA017CzXmp7dwS15gjHmGtaYQ4M47T2hiA6ak54nQesKchxVvSvwQEwQx7p8gaP7AA5GwDaEfXNDQ0BdloHiGVs5NWCDCRIMxtDQhoAQrmPCpFB62DcETCxilQKoACbEKTmKUYWgELrFTroKhwdQCJNwMk8YtDaq5wavnZhbVi6l26lwx0m0FIFHkbCLJYQkTjEukEdQ0wxrVA0PSJQL01CtO0PUnJEwwjujzEvCAmTV4RF9NAeYOgFI+QmWQKZDV5hcAUpQRZo0kDyCsFJRYuBigP0ydULAbIhwDOICkihjAXK8yUgLCeTAhzCguY0sYqVcjPPEMA3RX86ifM2i84MgzUpDhsc9Q69jwGvPWHk28JNugZ1ZF3aO5cKlV2MrPNG4N54E1sYdUOPtcWWhtngZFsBQxAA)

_No response_

**Comment by Rec0iL99** [comment]
Hi @Dahaden, thanks for the issue. Have you tried catching this using the [`require-atomic-updates`](https://eslint.org/docs/latest/rules/require-atomic-updates) rule?

**Comment by Dahaden** [comment]
Ahh @Rec0iL99 Yes that looks like the same thing :) Sorry had no idea what it was called and rough searches for `async` or `await` didn't lead me there :)
Thanks!
